### PR TITLE
[Serve] Wait for request to start processing in `test_http_replica_gauge_metrics`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

`test_metrics.py` has been flaky lately:

Screen Shot 2023-07-06 at 11 52 55 AM

See [example traceback](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4818#_):

```console
serve_start_shutdown = <ray.serve._private.client.ServeControllerClient object at 0x7f3088011d00>

    def test_http_replica_gauge_metrics(serve_start_shutdown):
        """Test http replica gauge metrics"""
        signal = SignalActor.remote()
    
        @serve.deployment
        class A:
            async def __call__(self):
                await signal.wait.remote()
    
        handle = serve.run(A.bind(), name="app1")
        _ = handle.remote()
        wait_for_condition(
            lambda: len(get_metric_dictionaries("serve_replica_processing_queries")) == 1,
            timeout=20,
        )
        processing_requests = get_metric_dictionaries("serve_replica_processing_queries")
        assert len(processing_requests) == 1
        assert processing_requests[0]["deployment"] == "app1_A"
        assert processing_requests[0]["application"] == "app1"
        print("processing_requests working as expected.")
    
        pending_requests = get_metric_dictionaries("serve_replica_pending_queries")
        assert len(pending_requests) == 1
        assert pending_requests[0]["deployment"] == "app1_A"
        assert pending_requests[0]["application"] == "app1"
    
        resp = requests.get("http://127.0.0.1:9999").text
        resp = resp.split("\n")
        for metrics in resp:
            if "# HELP" in metrics or "# TYPE" in metrics:
                continue
            if "serve_replica_processing_queries" in metrics:
>               assert "1.0" in metrics
E               assert '1.0' in 'ray_serve_replica_processing_queries{Component="core_worker",NodeAddress="172.16.16.3",SessionName="session_2023-07-05_17-34-28_672200_4914",Version="3.0.0.dev0",WorkerId="fe8d8074e49baebaf85a11429a49a6f019fd9194e2f7c047e400e545",application="app1",deployment="app1_A",replica="app1_A#ogIsqV"} 0.0'

python/ray/serve/tests/test_metrics.py:123: AssertionError
```

The root cause seems to be that the request occasionally takes some time stop pending and start processing. This change places that check inside a `wait_for_condition` call to remove flakiness.

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change fixes `test_metrics.py`.

### CODE REVIEW

File: python/ray/serve/tests/test_metrics.py [hunk 1]
```diff
@@ -92,37 +92,44 @@ def test_http_replica_gauge_metrics(serve_start_shutdown):
...
+        resp = resp.split("\n")
+        for metrics in resp:
+            if "# HELP" in metrics or "# TYPE" in metrics:
+                continue
+            if "serve_replica_processing_queries" in metrics:
+                print(metrics)
+                assert "1.0" in metrics
+            elif "serve_replica_pending_queries" in metrics:
+                assert "0.0" in metrics
+        return True
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_metrics.py:136]: > `ensure_request_processing()` will error when the condition are not met. And if all conditions are met we that will always be success case?

That's right.

> Should we maybe also check to ensure `serve_replica_processing_queries` and `serve_replica_pending_queries` exist in metrics?

Yes, that's a good idea. I added a check to make sure that both metrics exist.

[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/tests/test_metrics.py:136]: Cool thanks Shreyas!

**Review by GeneDer — APPROVED** [review]
:shipit:

**Comment by shrekris-anyscale** [comment]
I ran `test_metrics` 6 times (twice for each suite of Serve tests), and it didn't flake in any of them. This change is ready to merge @edoakes @architkulkarni.

**Comment by edoakes** [comment]
Good stuff!

**Pull request merged**